### PR TITLE
[TT-12885] Add plugin development guide for manual builds

### PR DESCRIPTION
**Pull request by titpetric**
### **User description**
<details open>
  <summary><a href="https://tyktech.atlassian.net/browse/TT-12885" title="TT-12885" target="_blank">TT-12885</a></summary>
  <br />
  <table>
    <tr>
      <th>Summary</th>
      <td>[Plugin docs] Provide guidance on compiling plugins manually</td>
    </tr>
    <tr>
      <th>Type</th>
      <td>
        Story
        Story
      </td>
    </tr>
    <tr>
      <th>Status</th>
      <td>In Dev</td>
    </tr>
    <tr>
      <th>Points</th>
      <td>N/A</td>
    </tr>
    <tr>
      <th>Labels</th>
      <td>-</td>
    </tr>
  </table>
</details>

---

https://tyktech.atlassian.net/browse/TT-12885

___

### **PR Type**
Documentation, Enhancement

___

### **Description**
- Added a comprehensive guide for Go plugin development, detailing environment setup, workspace creation, and plugin testing.
- Introduced Taskfiles for automating markdown formatting and plugin development tasks, enhancing workflow efficiency.
- Fixed indentation issues in the swagger documentation to improve readability.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>Taskfile.yml</strong><dd><code>Add Taskfile for markdown formatting with mdox</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

docs/Taskfile.yml

<li>Added a Taskfile for managing markdown formatting tasks.<br> <li> Introduced a default task to run <code>mdox</code> for formatting markdown files.

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6598/files#diff-490292b48d08097a4846f074704602500f054c4b9272545224d27b5ad0e24ca0">+8/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>Taskfile.yml</strong><dd><code>Add Taskfile for automating plugin development tasks</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

docs/plugins/Taskfile.yml

<li>Added a Taskfile for automating plugin development tasks.<br> <li> Defined tasks for checking out repositories, creating workspaces, and <br>testing plugins.

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6598/files#diff-d9772a9122a4c0e8eeb789a79dbe91a293a2f6b6ca72b43fbb73c62eef30d7de">+61/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>README.md</strong><dd><code>Add comprehensive Go plugin development guide</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

docs/plugins/README.md

<li>Added a comprehensive guide for Go plugin development.<br> <li> Included steps for setting up the environment, creating workspaces, <br>and testing plugins.<br> <li> Provided troubleshooting tips for common issues and debugging.

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6598/files#diff-aae529ed8d07326b707ba5a2998f0f5c0dea06651de706ea951c105d966417e2">+245/-0</a>&nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Formatting</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>swagger.md</strong><dd><code>Fix indentation in swagger generation documentation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

docs/swagger.md

<li>Fixed indentation issues in the documentation for generating swagger <br>files.

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6598/files#diff-0d3265e8be331704b75414619e8cf0fcf63cdce0bcb13ae1df64cf99e8e42b20">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

**Comment by buger** [comment]
<p>
    Let's make that PR title a 💯 shall we? 💪
    </p>
    <p>
    Your <em>PR title</em> and <em>story title</em> look <strong>slightly different</strong>. Just checking in to know if it was intentional!
    </p>
    <table>
      <tr>
        <th>Story Title</th>
        <td>[Plugin docs] Provide guidance on compiling plugins manually</td>
      </tr>
      <tr>
          <th>PR Title</th>
          <td>[TT-12885] Add plugin development guide for manual builds</td>
        </tr>
    </table>
    <p>
      Check out this <a href="https://www.atlassian.com/blog/git/written-unwritten-guide-pull-requests">guide</a> to learn more about PR best-practices.
    </p>

File: docs/plugins/README.md [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+task: [workspace:plugins] go mod init testplugin
+go: creating new go.mod: module testplugin
+task: [workspace:plugins] go mod edit -go "1.22.6"
+task: [workspace:plugins] wget -q https://raw.githubusercontent.com/TykTechnologies/custom-go-plugin/refs/heads/master/go/src/CustomGoPlugin.go
+```
+
+The internal `workspace:plugins` step ensures a few things:
+
+1. create a plugin, create go.mod,
+2. set go.mod go version to what's set in gateway,
```
Comment (lghiur) [inline docs/plugins/README.md:61]: can you place a link to GW go.mod so the user doesn't need to click around?

File: docs/plugins/README.md [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+3. have some code to compile in the folder
```
Comment (lghiur) [inline docs/plugins/README.md:62]: I would reorganise a bit these steps:

## Create a plugin:
- create main.go (or what is the name of the file) with your plugin code logic
- create go.mod
- set go.mod to Gateway go.mod version (link to GW go.mod)

File: docs/plugins/README.md [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+
+At this point, we don't have a workspace yet.
```
Comment (lghiur) [inline docs/plugins/README.md:64]: does it require more info on what a workspace is? or why is it needed.

"In order to be able to share dependencies across the go modules, you also need to create a **workspace**, So let's do that as a next step" - smth like this

**Comment by lghiur** [comment]
@titpetric I would propose to build a plugin together with the steps, like a hello world plugin, that they can copy paste from the guidance and endup with a functional plugin

**Comment by lghiur** [comment]
> @titpetric I would propose to build a plugin together with the steps, like a hello world plugin, that they can copy paste from the guidance and endup with a functional plugin

I see that it uses CustomGoPlugin in the example, which is perfect.

**Review by lghiur — APPROVED** [review]

**Review by andyo-tyk — APPROVED** [review]
On the assumption that all agreed changes from Tyk Docs review have been applied, I'm happy to approve.

**Pull request merged**